### ISSUE TITLE
2.6: vmap doesn't vectorize all args

### ISSUE THREAD

**Issue by Birch-san** [pytorch/pytorch#151561]
### 🐛 Describe the bug

vmap unbinds batch dimension for my first argument x, but not for my subsequent argument `mask`.  
how do I give a corresponding mask to each ensemble member?

I mean, strictly speaking for my use-case, my mask is the same for all ensemble members. if that worked I'd be happy too. but SDPA's validation of masks' batch dimensions is broken when under vmap (https://github.com/pytorch/pytorch/issues/151558).

```python
from typing import Optional
import torch
from torch import BoolTensor, FloatTensor, Tensor, inference_mode
from torch.nn import Module
from torch.func import functional_call, stack_module_state

class Printer(Module):
    def forward(self, x: FloatTensor, mask: Optional[BoolTensor] = None) -> FloatTensor:
        print(f"x: {x.shape}")
        if mask is not None:
            print(f"mask: {mask.shape}")
        # vmap is required to return a tensor
        return x

ensemble_size = 3
stateless, *experts = [Printer() for _ in range(ensemble_size + 1)]
params, buffers = stack_module_state(experts)

def bound_vmappable_fwd(
    params: dict[str, Tensor],
    buffers: dict[str, Tensor],
    x: FloatTensor,
    mask: Optional[BoolTensor] = None,
) -> FloatTensor:
    return functional_call(
        stateless,
        (params, buffers),
        (x, mask),
        tie_weights=False,
        strict=True,
    )

def bound_vmappable_fwd_kwarg(
    params: dict[str, Tensor],
    buffers: dict[str, Tensor],
    x: FloatTensor,
    mask: Optional[BoolTensor] = None,
) -> FloatTensor:
    return functional_call(
        stateless,
        (params, buffers),
        (x,),
        {'mask': mask},
        tie_weights=False,
        strict=True,
    )

device = torch.device('cuda')
batch_size = 4
seq_len = 128
in_features = 320
with inference_mode():
    x = torch.randn((batch_size, seq_len, in_features), device=device, dtype=torch.float16)
    mask = torch.ones((batch_size, 1, seq_len, seq_len), device=device, dtype=torch.bool)

    print("==running without vmap==")
    # prints:
    # x: torch.Size([4, 128, 320])
    # mask: torch.Size([4, 1, 128, 128])
    experts[0](x, mask=mask)

    print("==running with vmap==")
    # prints:
    # x: torch.Size([128, 320])
    # mask: torch.Size([4, 1, 128, 128])
    torch.vmap(bound_vmappable_fwd)(params, buffers, x, mask=mask)

    print("==running with vmap (kwargs)==")
    # prints:
    # x: torch.Size([128, 320])
    # mask: torch.Size([4, 1, 128, 128])
    torch.vmap(bound_vmappable_fwd_kwarg)(params, buffers, x, mask=mask)

    # it makes sense to me that vmap unbinds x's batch dim and gives a different x to each expert.
    # I'm surprised that it didn't explode when the batch dim didn't match the size of the ensemble. is that a bug too?
    # but why doesn't vmap unbind my mask's batch dimension and dispatch to each ensemble member?
```

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.13-650-3434-22042-coreweave-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3

Nvidia driver version: 535.216.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8462Y+
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           8
CPU max MHz:                        4100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5600.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hfi vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          3 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           128 MiB (64 instances)
L3 cache:                           120 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110,112,114,116,118,120,122,124,126
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95,97,99,101,103,105,107,109,111,113,115,117,119,121,123,125,127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] clip-anytorch==2.6.0
[pip3] dctorch==0.1.2
[pip3] DISTS-pytorch==0.1
[pip3] lovely-numpy==0.2.13
[pip3] mypy-extensions==1.0.0
[pip3] natten==0.17.4+torch260cu128
[pip3] numpy==1.26.4
[pip3] open_clip_torch==2.31.0
[pip3] optree==0.14.1
[pip3] torch==2.6.0
[pip3] torchaudio==2.6.0
[pip3] torchdiffeq==0.2.5
[pip3] torchsde==0.2.6
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0+git35c6c7c6
[pip3] welford-torch==0.2.5
[conda] Could not collect
```

**Comment by Birch-san** [comment]
ah, looks like my attempt to check positional args was incorrect. this does actually work.

```python
from typing import Optional
import torch
from torch import BoolTensor, FloatTensor, Tensor, inference_mode
from torch.nn import Module
from torch.func import functional_call, stack_module_state

class Printer(Module):
    def forward(self, x: FloatTensor, mask: Optional[BoolTensor] = None) -> FloatTensor:
        print(f"x: {x.shape}")
        if mask is not None:
            print(f"mask: {mask.shape}")
        # vmap is required to return a tensor
        return x

ensemble_size = 3
stateless, *experts = [Printer() for _ in range(ensemble_size + 1)]
params, buffers = stack_module_state(experts)

def bound_vmappable_fwd(
    params: dict[str, Tensor],
    buffers: dict[str, Tensor],
    x: FloatTensor,
    mask: Optional[BoolTensor] = None,
) -> FloatTensor:
    return functional_call(
        stateless,
        (params, buffers),
        (x, mask),
        tie_weights=False,
        strict=True,
    )

device = torch.device('cuda')
batch_size = 4
seq_len = 128
in_features = 320
with inference_mode():
    x = torch.randn((batch_size, seq_len, in_features), device=device, dtype=torch.float16)
    mask = torch.ones((batch_size, 1, seq_len, seq_len), device=device, dtype=torch.bool)

    print("==running without vmap==")
    # prints:
    # x: torch.Size([4, 128, 320])
    # mask: torch.Size([4, 1, 128, 128])
    experts[0](x, mask=mask)

    print("==running with vmap (positional args) ==")
    # prints:
    # x: torch.Size([128, 320])
    # mask: torch.Size([1, 128, 128])
    torch.vmap(bound_vmappable_fwd)(params, buffers, x, mask)

    print("==running with vmap (kwargs)==")
    # prints:
    # x: torch.Size([128, 320])
    # mask: torch.Size([4, 1, 128, 128])
    torch.vmap(bound_vmappable_fwd)(params, buffers, x, mask=mask)
```

the reason for my surprise is that the torch.vmap docs describe the returned function as follows:

> Returns a new “batched” function. It takes the same inputs as func, except each input has an extra dimension at the index specified by in_dims.

https://pytorch.org/functorch/stable/generated/functorch.vmap.html  
https://github.com/pytorch/pytorch/blob/d8bafd23ab2cd3c1edb94841b14d075b87ed78a8/torch/_functorch/apis.py#L31

however it looks like this caveat _is_ mentioned lower down. it's a bit buried:

> For any function that uses kwargs, the returned function will not batch the kwargs but will accept kwargs

so, the only action I can think of is for docs to signpost this caveat better (i.e. don't describe the return type as "all inputs are batched").

**Comment by zou3519** [comment]
Just to check: if you pass mask as a kwarg (e.g. mask=mask), then vmap should not batch it. However, if you pass mask as a positional arg, I would expect vmap to batch it. Does that happen?

**Comment by Birch-san** [comment]
yes, that’s what I said above. I provided a repro too. I also explained where the docs contradict themselves, and proposed how they should be changed.
